packages/fxa-auth-db-mysql/lib/db/schema/patch-108-109.sql

```diff
@@ -0,0 +1,97 @@
+-- This migration will help us to abstract the half baked recovery keys of fxa accounts. In this migration,
+-- we're adding a new `upsertRecoveryKey` procedure that will handle the half baked recovery keys by using
+-- the following criterias.
+
+-- 1. If a user has no recovery key then create a new recovery key.
+-- 2. If a user has unverified recovery key then delete that recovery key and create a new unverified
+--    recovery key.
+-- 3. If a user has verified recovery key then throw an exception
+
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+CALL assertPatchLevel('108');
+
+-- Delete these below 3 lines
+DROP PROCEDURE IF EXISTS createRecoveryKey_5;
+DROP PROCEDURE IF EXISTS getRecoveryKey_5;
+DROP PROCEDURE IF EXISTS upsertRecoveryKey_1;
+
+CREATE PROCEDURE `createRecoveryKey_5` (
+  IN `uidArg` BINARY(16),
+  IN `recoveryKeyIdHashArg` BINARY(32),
+  IN `recoveryDataArg` TEXT,
+  IN `createdAtArg` BIGINT UNSIGNED,
+  IN `enabledArg` BOOLEAN
+)
+BEGIN
+
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  SET @accountCount = 0;
+
+  -- Signal error if no user found
+  SELECT COUNT(*) INTO @accountCount FROM `accounts` WHERE `uid` = `uidArg`;
+  IF @accountCount = 0 THEN
+    SIGNAL SQLSTATE '45000' SET MYSQL_ERRNO = 1643, MESSAGE_TEXT = 'Can not create recovery key for unknown user.';
+  END IF;
+
+  INSERT INTO recoveryKeys (
+    uid,
+    recoveryKeyIdHash,
+    recoveryData,
+    createdAt,
+    enabled
+  )
+  VALUES (
+    uidArg,
+    recoveryKeyIdHashArg,
+    recoveryDataArg,
+    createdAtArg,
+    enabledArg
+  );
+
+  UPDATE accounts SET profileChangedAt = (UNIX_TIMESTAMP(NOW(3)) * 1000) WHERE uid = uidArg;
+
+  COMMIT;
+END;
+
+CREATE PROCEDURE `upsertRecoveryKey_1` (
+  IN `uidArg` BINARY(16),
+  IN `recoveryKeyIdHashArg` BINARY(32),
+  IN `recoveryDataArg` TEXT,
+  IN `createdAtArg` BIGINT UNSIGNED,
+  IN `enabledArg` BOOLEAN
+)
+BEGIN
+
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    RESIGNAL;
+  END;
+
+  SET @recoveryKeyCount = NULL;
```

Comment:
I think another way we can look at this that might make the logic easier to follow is this:

1. Does user have a verified recovery key? If so throw error
2. Delete all unverified recovery keys
3. Create recovery key